designs/enum-shapes.md

```diff
@@ -0,0 +1,259 @@
+# Enum shapes in Smithy
+
+* **Author**: Michael Dowling
+* **Created**: 2022-02-14
+* **Last updated**: 2022-02-14
+
+## Abstract
+
+This proposal introduces an `enum` shape and `intEnum` shape to Smithy to
+expand the existing `@enum` trait capabilities.
+
+## Motivation
+
+There are two primary motivations:
+
+1. Smithy does not provide the ability to define an enumerated integer, and
+   many serialization formats serialize enums as integers rather than strings.
+2. Enum values are similar to members, but defined in a less flexible way that
+   requires special casing. Enums in Smithy have enum values that are very
+   similar to members, but cannot be targeted by traits or filtered like other
+   members. For example, an `@enum` trait supports the value, name,
+   documentation, tags, and deprecated properties, of which documentation,
+   tags, and deprecated provide functionality that is exactly equivalent to
+   similarly named traits. In order to filter enums out of a model, model
+   transformations need to write special-cased code to deal with enum traits
+   rather than relying on the existing logic used to remove member shapes.
+
+## Proposal
+
+Two new simple shapes will be introduced in Smithy 2.0:
+
+* `enum`: An enumerated type that is represented using string values
+* `intEnum`: An enumerated type that is represented using integer values.
+
+### Enum shape
+
+The enum shape is used to represent a fixed set of string values. The following
+example defines an enum shape:
+
+```
+enum Suit {
+    DIAMOND
+    CLUB
+    HEART
+    SPADE
+}
+```
+
+Each value listed in the enum is a member that implicitly targets
+`smithy.api#Unit`. The string representation of an enum member defaults to the
+member name. The string representation can be customized by applying the
+`@enumValue` trait.
+
+```
+enum Suit {
+    @enumValue(string: "diamond")
+    DIAMOND
+
+    @enumValue(string: "club")
+    CLUB
+    
+    @enumValue(string: "heart")
+    HEART
+    
+    @enumValue(string: "spade")
+    SPADE
+}
+```
+
+Enums do not support aliasing; all values MUST be unique.
+
+#### enum default values
+
+The default value of the enum shape is an empty string "", regardless of if
+the enum defines a member with a value of "".
+
+```
+structure GetFooInput {
+    @default
+    suit: Suit
+}
+```
+
+To account for this, enums MAY define a default member by making the
+`@enumValue` of the member "":
+
+```
+enum Suit {
+    @enumValue(string: "")
```

Comment:
Are you suggesting we disallow setting those values explicitly?

Reply:
Yes, mostly because of how `0` has a big impact on `intEnum` when used with the `@default` trait. People might not realize when setting the value to `0` that they are also defining the default value of the enum. By adding a new property or trait, they have to realize it because it's on the tin.